src/D2L.CodeStyle.Analyzers/ApiUsage/DataHub/ExternalDataSetVisibilityProviderAnalyzer.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage.DataHub {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal class ExternalDataSetVisibilityProviderAnalyzer : DiagnosticAnalyzer {
+
+		public const string IEventDrivenDataSetPluginFullName = "D2L.AW.DataExport.BrightspaceDataSets.Domain.IEventDrivenDataSetPlugin";
+		public const string ExternalDataSetVisibilityProviderFullName = "D2L.AW.DataExport.BrightspaceDataSets.Domain.IExternalDataSetVisibilityProvider`1";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+				Diagnostics.ExternalDataSetVisibilityProviderTypeParameterMatchesClass
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			INamedTypeSymbol IEventDrivenDataSetPluginType = compilation.GetTypeByMetadataName( IEventDrivenDataSetPluginFullName );
+			if( IEventDrivenDataSetPluginType.IsNullOrErrorType() ) {
+				return;
+			}
+
+			INamedTypeSymbol ExternalDataSetVisibilityProviderType = compilation.GetTypeByMetadataName( ExternalDataSetVisibilityProviderFullName );
+			if( ExternalDataSetVisibilityProviderType.IsNullOrErrorType() ) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(

```

Comment:
Doing this as a syntax action means it might be called multiple times for partial classes:

```cs
// here
partial class Foo {}

// and here
partial class Foo {}
```

You might want `RegisterSymbolAction` or `RegisterSymbolStartAction`.